File: Olympiad/POI/official/04-XI/zga/dla_zaw/zga_pas/kasyno.pas
uses zga;

const
  PREC = 12;
  ILE_ZESTAWOW = 1000000;
  PUSTE = -1;

  WYPLATA = 423.99;
  DOPLATA = 13.53;

function daj_losowa : Double;
var
  i : Integer;
  x : Double;
begin
  x := 0.0;
  repeat
    for i := 1 to PREC do
      x := (x + random (10)) / 10;
  until x <> 0.0;
  daj_losowa := x;
end;

var
  wygrane : LongInt;
  i : LongInt;
  j, k, pozycja : Integer;
  t : array [0..ILE_LICZB-1] of Double;
  s : array [0..ILE_LICZB] of Double;
  wygrana, powtorzyl : Boolean;
  pkt : Integer;

begin
  wygrane := 0;

  Randomize;

  inicjalizuj;

  for i := 1 to ILE_ZESTAWOW do
  begin
    nowa_rozgrywka;

    s [0] := 0;
    for j := 1 to ILE_LICZB do
      s [j] := PUSTE;

    for j := 0 to ILE_LICZB - 1 do
    begin
      repeat
        t[j] := daj_losowa;
        powtorzyl := False;
        for k := 0 to j - 1 do
          if t [j] = t [k] then
            powtorzyl := True;
      until not powtorzyl;

      pozycja := kolejna_liczba (t [j]);
      if (pozycja >= 1) and (pozycja <= ILE_LICZB) then
        s [pozycja] := t [j];
    end;

    //sprawdzanie, czy tablica s jest posortowana
    wygrana := True;
    for j :=1 to ILE_LICZB do
      if (s[j] = PUSTE) or (s [j] <= s [j-1]) then
        wygrana := False;

    if wygrana then
      Inc (wygrane);

  end;

  pkt := Round (1e-4 * (wygrane * WYPLATA - (ILE_ZESTAWOW - wygrane) * DOPLATA));

  if pkt < 0 then pkt := 0;
  if pkt > 100 then pkt := 100;

  WriteLn ('Liczba wygranych: ', wygrane);
  WriteLn ('Liczba punktów: ', pkt);

end.

File: Olympiad/POI/official/04-XI/zga/kasyno.c
#ifdef __cpluslus
#include <cstdio>
#include <cstdlib>
#else
#include <stdio.h>
#include <stdlib.h>
#endif

#include "zga.h"

#define PREC 12
#define ILE_ZESTAWOW 1000000L
#define PUSTE -1.0

#define MIN_LICZB 2

#define WYPLATA 423.99
#define DOPLATA 13.53

double daj_losowa()
{
  int i;
  double x = 0.0;
  do
/*    for (i=0; i++ < PREC; )
      x = (x + (int) (10.0*rand()/(RAND_MAX+1.0))) / 10;*/
    for (i=0; i < 2; i++)
      x = (x + rand() % 1000000) / 1000000;
  while (x == 0.0);
  return x;
}

int main ()
{
  long int wygrane = 0;
  long int i;
  int j, k, powtorzyl, wygrana, pozycja;
  double t [ILE_LICZB], s [ILE_LICZB + 1];
  double pkt_d;
  int pkt;

  srandom (5);

  inicjalizuj();

  for (i=0; i < ILE_ZESTAWOW; i++)
  {
    nowa_rozgrywka();

    s [0] = 0;
    for (j=1; j <= ILE_LICZB; j++)
      s [j] = PUSTE;

    for (j=0; j < ILE_LICZB; j++)
    {
      do
      {
        t[j] = daj_losowa ();
        powtorzyl = 0;
        for (k=0; k < j; k++)
          if (t [j] == t [k])
            powtorzyl = 1;
      }
      while (powtorzyl);

      pozycja = kolejna_liczba (t [j]);
      if (pozycja >= 1 && pozycja <= ILE_LICZB)
        s [pozycja] = t [j];
    }

    //sprawdzanie, czy tablica s jest posortowana
    wygrana = 1;
    for (j=1; j <= ILE_LICZB; j++)
      if (s[j] == PUSTE || s [j] <= s [j-1])
        wygrana = 0;

    if (wygrana)
      wygrane++;

  }

  pkt_d = 1e-4 * (((double)wygrane) * WYPLATA - ((double)(ILE_ZESTAWOW - wygrane)) * DOPLATA);
  pkt = pkt_d - (double)((int) pkt_d) < 0.5 ? (int) pkt_d : (int) pkt_d + 1; //zaokraglenie

  if (pkt < 0) pkt = 0;
  if (pkt > 100) pkt = 100;

  printf ("OK\n");
  printf ("%ld\n", wygrane);
  printf ("%d\n", pkt);

  return 0;
}
